ckanext/s3filestore/uploader.py

```diff
@@ -628,15 +627,21 @@ def update_visibility(self, id, target_acl=None):
             return
 
         for upload in resource_objects['Contents']:
-            is_public_read = self.is_key_public(upload['Key'])
+            upload_key = upload['Key']
+            log.debug("Setting visibility for key [%s], current object is [%s]", upload_key, current_key)
+            if upload_key == current_key:
+                acl = target_acl
+            else:
+                acl = 'private'
+            is_public_read = self.is_key_public(upload_key)
             # if the ACL status doesn't match what we want, update it
-            if (target_acl == PUBLIC_ACL) != is_public_read:
-                log.debug("Updating ACL for object %s to %s", upload['Key'], target_acl)
+            if (acl == PUBLIC_ACL) != is_public_read:
+                log.debug("Updating ACL for object %s to %s", upload_key, acl)
                 client.put_object_acl(
-                    Bucket=self.bucket_name, Key=upload['Key'], ACL=target_acl)
-                self._cache_delete(upload['Key'])
-                self._cache_put(_get_visibility_cache_key(upload['Key']), target_acl)
-        self._cache_put(_get_visibility_cache_key(id + '/all'), target_acl)
+                    Bucket=self.bucket_name, Key=upload_key, ACL=acl)
+                self._cache_delete(upload_key)
+                self._cache_put(upload_key + VISIBILITY_CACHE_PATH, acl)
+        self._cache_put(current_key + VISIBILITY_CACHE_PATH + '/all', target_acl)
```

Comment:
is it possible to set a longer window for /all and for /visibility keys, 1 hour may not be long enough for the performance improvements and with these items being so small, i think 30 days should be ok, similar for public files, they should be left in cache far longer than when they are private and expire.

Reply:
eg update signature of _cache_put to
def _cache_put(self, key, value, ttl=self.signed_url_cache_window):
so default is from cache_window but for acl targets, they can be 30 days for max speed and reduce s3 'hits'

Reply:
Hmm...since the cache is automatically cleared when a resource is uploaded, I suppose it is reasonable to have a fairly long expiry on it if that hasn't happened.